spacy_experimental/coref/span_resolver_component.py

```diff
@@ -0,0 +1,403 @@
+from typing import Iterable, Tuple, Optional, Dict, Callable, Any, List
+import warnings
+
+from thinc.types import Floats2d, Floats3d, Ints2d
+from thinc.api import Model, Config, Optimizer, CategoricalCrossentropy
+from thinc.api import set_dropout_rate, to_categorical
+from itertools import islice
+import srsly
+
+from spacy.pipeline.trainable_pipe import TrainablePipe
+from spacy.language import Language
+from spacy.training import Example, validate_examples, validate_get_examples
+from spacy.errors import Errors
+from spacy.tokens import Doc
+from spacy.vocab import Vocab
+from spacy.util import from_bytes, from_disk
+
+from .coref_util import (
+    MentionClusters,
+    DEFAULT_CLUSTER_PREFIX,
+)
+
+from .coref_scorer import doc2clusters, score_span_predictions
+
+default_span_resolver_config = """
+[model]
+@architectures = "spacy-experimental.SpanResolver.v1"
+hidden_size = 1024
+distance_embedding_size = 64
+conv_channels = 4
+window_size = 1
+max_distance = 128
+prefix = "coref_head_clusters"
+
+[model.tok2vec]
+@architectures = "spacy.Tok2Vec.v2"
+
+[model.tok2vec.embed]
+@architectures = "spacy.MultiHashEmbed.v1"
+width = 64
+rows = [2000, 2000, 1000, 1000, 1000, 1000]
+attrs = ["ORTH", "LOWER", "PREFIX", "SUFFIX", "SHAPE", "ID"]
+include_static_vectors = false
+
+[model.tok2vec.encode]
+@architectures = "spacy.MaxoutWindowEncoder.v2"
+width = ${model.tok2vec.embed.width}
+window_size = 1
+maxout_pieces = 3
+depth = 2
+"""
+DEFAULT_SPAN_RESOLVER_MODEL = Config().from_str(default_span_resolver_config)["model"]
+
+
+def span_resolver_scorer(examples: Iterable[Example], **kwargs) -> Dict[str, Any]:
+    return score_span_predictions(examples, **kwargs)
+
+
+def make_span_resolver_scorer():
+    return span_resolver_scorer
+
+
+@Language.factory(
+    "experimental_span_resolver",
+    assigns=["doc.spans"],
+    requires=["doc.spans"],
+    default_config={
+        "model": DEFAULT_SPAN_RESOLVER_MODEL,
+        "input_prefix": "coref_head_clusters",
+        "output_prefix": "coref_clusters",
+        "scorer": {"@scorers": "spacy-experimental.span_resolver_scorer.v1"},
+    },
+    default_score_weights={"span_accuracy": 1.0},
+)
+def make_span_resolver(
+    nlp: Language,
+    name: str,
+    model,
+    input_prefix: str = "coref_head_clusters",
+    output_prefix: str = "coref_clusters",
```

Comment:
Should we store these in global vars as well? Because they seem to get repeated a bit - also in the test suite.

Reply:
Good point, moved to vars in `coref_util`.